tidy3d/components/data/sim_data.py

```diff
@@ -230,23 +230,37 @@ def _at_centers(self, monitor_data: xr.Dataset) -> xr.Dataset:
         Returns
         -------
         xarray.Dataset
-            Dataset containing all of the fields in the data
-            interpolated to center locations on Yee grid.
+            Dataset containing all of the fields in the data interpolated to boundary locations on
+            the Yee grid.
         """
 
-        # discretize the monitor and get center locations
-        sub_grid = self.simulation.discretize(monitor_data.monitor, extend=False)
-        centers = sub_grid.centers
+        if monitor_data.monitor.colocate:
+            # TODO: this still errors if monitor_data.colocate is allowed to be ``True`` in the
```

Comment:
Yeah it's probably some sort of casting. However re adding a test, this PR actually adds a validator that `colocate=False` is set for `output_monitors`, because the adjoint source is not correct otherwise. So this comment is more like if in the future `colocate=True` is allowed (which may never happen?). Until then though a unit test doesn't really work as the `colocate=False` validator would have to be turned off. Thoughts?

Reply:
ok so we dont need to worry about this now I guess. (no need for the test). Regarding colocate=True in adjoint, would be accurate if we added reverse interpolation in the (volumetric) CustomCurrentSource?

Reply:
No, you can't actually reverse interpolate back to the original data unambiguously. Some information is lost in the interpolation, the reverse interpolation has to make some assumptions on how to recover the data which are only approximate. The correct thing to do is to backprop through the interpolation. So for example this could work as:

- user makes monitor with colocate = True
- under the hood plugin uses a monitor with colocate = False (so that the adjoint source is correct), and when loading the results makes a colocated `FieldData` (tracked by Jax) to return to user which can then be further manipulated (also tracked by jax)

Then on the backwards pass the interpolation is backpropped through correctly. This can happen because forward data is stored, while reverse interp is agnostic to forward data.